### PR TITLE
Added unit tests and support for card authorize

### PR DESCRIPTION
**Pull request by JaisDK**

### CODE REVIEW

File: src/Action/Api/ConfirmPaymentAction.php [hunk 1]
```diff
@@ -38,7 +38,7 @@ public function execute($request)
         $latestOperation = $quickpayPayment->getLatestOperation();
 
         if (null === $latestOperation) {
-            throw new LogicException('The payment does not have a `latest operation`');
+            throw new LogicException('The payment does not have a `latest operation`'); // @codeCoverageIgnore
```
Comment (loevgaard) [inline src/Action/Api/ConfirmPaymentAction.php]: We don't want to add these 'ignore' annotations just to get a higher coverage. It's untrue :D

File: src/Action/StatusAction.php [hunk 2]
```diff
@@ -71,7 +72,7 @@ public function execute($request): void
 
                 break;
             default:
-                $request->markUnknown();
+                $request->markUnknown(); // @codeCoverageIgnore
```
Comment (loevgaard) [inline src/Action/StatusAction.php]: Same here

File: src/Action/StatusAction.php [hunk 3]
```diff
@@ -31,7 +31,7 @@ public function execute($request): void
 
         $model = ArrayObject::ensureArrayObject($request->getModel());
 
-        if (!$model['quickpayPaymentId']) {
+        if (!$model['quickpayPaymentId'] && !$model['quickpayPayment']) {
```
Comment (loevgaard) [inline src/Action/StatusAction.php]: Would it be better to use `$model->offsetExists('quickpayPaymentId')`?

File: src/Api.php [hunk 4]
```diff
@@ -183,7 +211,7 @@ protected function doRequest($method, string $path, array $params = []): Respons
 
         $encodedParams = json_encode($params);
         if (false === $encodedParams) {
-            throw new \InvalidArgumentException('Could not encode $params');
+            throw new InvalidArgumentException('Could not encode $params'); // @codeCoverageIgnore
```
Comment (loevgaard) [inline src/Api.php]: Coverage annotation

File: src/Api.php [hunk 5]
```diff
@@ -234,9 +262,9 @@ public static function assertValidResponse(ResponseInterface $response, string $
     {
         if ($response->hasHeader('QuickPay-Checksum-Sha256')) {
             $checksum = self::checksum((string) $response->getBody(), $privateKey);
-            $qp_checksum = $response->getHeader('QuickPay-Checksum-Sha256');
+            $qp_checksum = $response->getHeaderLine('QuickPay-Checksum-Sha256');
```
Comment (loevgaard) [inline src/Api.php]: Camel case, not snake case

File: src/Model/QuickPayPayment.php [hunk 6]
```diff
@@ -14,6 +14,31 @@ class QuickPayPayment extends QuickPayModel
...
+
+    /**
+     * @var string
+     */
+    protected $currency;
+
+    /**
+     * string.
+     */
+    protected $order_id;
```
Comment (loevgaard) [inline src/Model/QuickPayPayment.php:30]: Camel case

File: src/Model/QuickPayPayment.php [hunk 7]
```diff
@@ -14,6 +14,31 @@ class QuickPayPayment extends QuickPayModel
...
+     */
+    protected $id;
+
+    /**
+     * @var string
+     */
+    protected $currency;
+
+    /**
+     * string.
```
Comment (loevgaard) [inline src/Model/QuickPayPayment.php]:
```suggestion
     * @var string
```

File: src/Model/QuickPayPaymentOperation.php [hunk 8]
```diff
@@ -8,9 +8,31 @@ class QuickPayPaymentOperation extends QuickPayModel
...
+
+    /**
+     * @var int
+     */
+    protected $amount;
+
+    /**
+     * @var string
+     */
+    protected $qp_status_code;
```
Comment (loevgaard) [inline src/Model/QuickPayPaymentOperation.php:34]:
```suggestion
    protected $quickpayStatusCode;
```

I would probably call it that

File: src/Model/QuickPayPaymentOperation.php [hunk 9]
```diff
@@ -41,31 +63,31 @@ public static function createFromArray(array $operations): array
...
+        return $this->type;
     }
 
     /**
      * @return int
      */
     public function getAmount(): int
     {
-        return $this->data->amount;
+        return $this->amount;
```
Comment (loevgaard) [inline src/Model/QuickPayPaymentOperation.php:82]: Why isn't this cast to int when the status code is?

File: src/Model/QuickpayCard.php [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+
+    /**
+     * @var string
+     */
+    protected $expiration;
+
+    /**
+     * @var int
+     */
+    protected $cvd;
```
Comment (loevgaard) [inline src/Model/QuickpayCard.php:22]: Is this the three digit number on a credit card? Is so, isn't is called `cvc`?

File: src/Api.php [hunk 11]
```diff
@@ -73,13 +73,34 @@ public function getPayment(ArrayObject $params, bool $create = true): QuickPayPa
...
+     *
+     * @return QuickPayPayment[]
+     */
+    public function getPayments(ArrayObject $params): array
+    {
+        $params = ArrayObject::ensureArrayObject($params);
+
+        $response = $this->doRequest('GET', 'payments?'.http_build_query($params->getArrayCopy()));
+
+        $payments = json_decode((string) $response->getBody());
```
Comment (loevgaard) [inline src/Api.php:94]: From the php docs:

> NULL is returned if the json cannot be decoded or if the encoded data is deeper than the recursion limit.

So we need to handle this case

File: src/Api.php (hunk 5 above)
Comment (loevgaard) [reply src/Api.php]:
```suggestion
            $quickpayChecksum = $response->getHeaderLine('QuickPay-Checksum-Sha256');
```

**Review by loevgaard — CHANGES_REQUESTED** [review]

File: src/Model/QuickPayPayment.php (hunk 6 above)
Comment (JaisDK) [reply src/Model/QuickPayPayment.php:30]: It has to match the keys from the QP response

File: src/Model/QuickPayPaymentOperation.php (hunk 9 above)
Comment (JaisDK) [reply src/Model/QuickPayPaymentOperation.php:82]: status code is a string in the response, amount is not

File: src/Model/QuickpayCard.php (hunk 10 above)
Comment (JaisDK) [reply src/Model/QuickpayCard.php:22]: it can be called both, QP use `cvd` as the key

File: src/Api.php [hunk 12]
```diff
@@ -73,13 +73,37 @@ public function getPayment(ArrayObject $params, bool $create = true): QuickPayPa
...
+        if (false === $payments) {
```
Comment (loevgaard) [inline src/Api.php]:
```suggestion
        if (null === $payments) {
```

**Review by loevgaard — CHANGES_REQUESTED** [review]

**Review by loevgaard — APPROVED** [review]

**Pull request merged**